text/0000-stabilize-marker-freeze.md

```diff
@@ -0,0 +1,149 @@
+
+- Some people have previously argued that this would be akin to exposing compiler internals.
+	- The RFC author disagrees, viewing `Freeze` in a similar light as `Send` and `Sync`: a trait that allows soundness requirements to be proven at compile time.
+- `Freeze` being an auto-trait, it is, like `Send` and `Sync` a sneaky SemVer hazard.
+	- Note that this SemVer hazard already exists through the existence of static-promotion, as exemplified by the following example:
+	```rust
+    // old version of the crate.
+    mod v1 {
+        pub struct S(i32);
+        impl S {
+            pub const fn new() -> Self { S(42) }
+        }
+    }
+
+    // new version of the crate, adding interior mutability.
+    mod v2 {
+        use std::cell::Cell;
+        pub struct S(Cell<i32>);
+        impl S {
+            pub const fn new() -> Self { S(Cell::new(42)) }
+        }
+    }
+
+    // Old version: builds
+    const C1: &v1::S = &v1::S::new();
+    // New version: does not build
+    const C2: &v2::S = &v2::S::new();
+	```
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+- The benefits of stabilizing `core::mem::Freeze` have been highlighted in [Motivation](#motivation).
+	- By not stabilizing `core::mem::Freeze` in trait bounds, we are preventing useful and sound code patterns from existing which were previously supported.
+- Alternatively, a non-auto sub-trait of `core::mem::Freeze` may be defined:
+	- While this reduces the SemVer hazard by making its breakage more obvious, this does lose part of the usefulness that `core::mem::Freeze` would provide to projects such as `zerocopy`.
+	- A "perfect" derive macro should then be introduced to ease the implementation of this trait. A lint may be introduced in `clippy` to inform users of the existence and applicability of this new trait.
+
+# Prior art
+[prior-art]: #prior-art
+
+- This feature has been available in `nightly` for 7 years, and is used internally by the compiler.
+- The work necessary for this RFC has already been done and merged in [this PR](https://github.com/rust-lang/rust/issues/121675), and a [tracking issue](https://github.com/rust-lang/rust/issues/121675) was opened.
+- zerocopy's [`Immutable`](https://docs.rs/zerocopy/0.8.0-alpha.11/zerocopy/trait.Immutable.html) seeks to provide the same guarantees as `core::marker::Freeze`.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+[Should the trait be exposed under a different name?](https://github.com/rust-lang/rust/pull/121501#issuecomment-1962900148)
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+- One might later consider whether `core::mem::Freeze` should be allowed to be `unsafe impl`'d like `Send` and `Sync` are, possibly allowing wrappers around interiorly mutable data to hide this interior mutability from constructs that require `Freeze` if the logic surrounding it guarantees that the interior mutability will never be used.
```

Comment:
I don't think this is a plausible possibility. If you never use interior mutability, just don't use a type with UnsafeCell. What is the possible use-case for having an UnsafeCell that you will *definitely never use for interior mutability*? (That would be the only situation where `unsafe impl Freeze` would be sound.)

Reply:
We've considered wanting this for zerocopy. The use case would be to add a wrapper type `Frozen<T>` (joking - no clue what we'd name it) that is `Freeze` even if `T` isn't. IIUC this would require `Freeze` to be a runtime property rather than a type property (ie, interior mutation never happens), and would require being able to write `unsafe impl<T> Freeze for Frozen<T>`.

Reply:
> What is the possible use-case for having an UnsafeCell that you will definitely never use for interior mutability?

Speculation: Perhaps a data structure whose contents are computed using interior mutability, then put in a wrapper struct which disables all mutation operations and implements `Freeze`. The reason to do this using a wrapper rather than transmuting to a non-interior-mutable type would be to avoid layout mismatches due to `UnsafeCell<T>` not having niches that `T` might have.

This is only relevant if such a type _also_ wants to support the functionality enabled by implementing `Freeze`, of course.

Reply:
> We've considered wanting this for zerocopy. The use case would be to add a wrapper type Frozen<T> (joking - no clue what we'd name it) that is Freeze even if T isn't.

What would that type be good for?

Reply:
> > We've considered wanting this for zerocopy. The use case would be to add a wrapper type Frozen (joking - no clue what we'd name it) that is Freeze even if T isn't.
> 
> What would that type be good for?

It's been a while since we considered this design, so the details are hazy in my memory, but roughly we wanted to be able to validate that a `&[u8]` contained the bytes of a bit-valid `&T`. We wanted to encode in the type system that we'd already validated size and alignment, so we wanted a type that represented "this has the size and alignment of `T` and all of its bytes are initialized, but might not be a bit-valid `T`." To do this, we experimented with a `MaybeValid<T>` wrapper.

When you're only considering by-value operations, you can just do `#[repr(transparent)] struct MaybeValid<T>(MaybeUninit<T>);` (`MaybeUninit` itself isn't good enough because it doesn't promise that its bytes are initialized - the newtype lets you add internal invariants). The problem is that this doesn't work by reference - if you're using `Freeze` as a bound to prove that `&[u8] -> &MaybeValid<T>` is sound, it won't work if `T: !Freeze`.

As I said, we didn't end up going that route - instead of constructing a `&MaybeValid<T>`, we construct (roughly) a `NonNull<T>` (okay, actually a [`Ptr`](https://google.github.io/zerocopy/zerocopy/pointer/ptr/def/struct.Ptr.html)) and just do the operations manually. The code in question starts at [this API](https://docs.rs/zerocopy/0.8.0-alpha.14/zerocopy/trait.TryFromBytes.html#method.try_ref_from) if you're curious.

Reply:
If you're wrapping `T` inside a private field, why does it matter whether it has interior mutability or not?

The only possible benefit I can see here would be still getting `noalias` optimizations for `Frozen<T>`, but you're not mentioning optimizations. In terms of soundness, if making things less `Freeze` can never make code unsound. (https://github.com/rust-lang/unsafe-code-guidelines/issues/303 is a Stacked Borrows bug.)

Reply:
> If you're wrapping `T` inside a private field, why does it matter whether it has interior mutability or not?

Because we want to be able to use `MaybeValid<T>` with APIs that are safe and use trait bounds (like `Freeze`) to prove their soundness internally. One of our golden rules for developing zerocopy is to be extremely anal about *internal* abstraction boundaries [1]. While we could just do what you're suggesting using `unsafe` (namely, use `unsafe` to directly do the `&[u8] -> &MaybeValid<T>` cast), we have existing internal APIs such as [this one](https://google.github.io/zerocopy/zerocopy/pointer/ptr/def/struct.Ptr.html#method.try_cast_into) that permit this conversion safely (ie, the method itself is safe to call), using trait bounds to enforce soundness.

In general, we've found that, as we teach our APIs to handle more and more cases, it quickly becomes very unwieldy to program "directly" using one big `unsafe` block (or a sequence of `unsafe` blocks) in each API. Decomposition using abstractions like `Ptr` has been crucial to us being confident that the code we're writing is actually sound.

[1] See "Abstraction boundaries and unsafe code" in our [contributing guidelines](https://github.com/google/zerocopy/discussions/1318)

Reply:
I'm afraid it is not clear to me how `Frozen` helps build up internal abstraction barriers, and your existing codebase is way too big for me to be able to extract that kind of information from it with reasonable effort.

Reply:
*TLDR: We can only make this code compile if `MaybeValid<T>: Freeze` despite `T: ?Freeze`. I'm arguing that, in order to support this use case, we should design the RFC to be forwards-compatible with permitting unsafe `impl<T> Freeze for MaybeValid<T>` even if we don't support it out of the gate.*

[Here's a minification](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=ce0b446f00f2fad293f4f1c8e64d40ba) of the sort of code I'm describing. It has three components that, in reality, would live in unrelated parts of our codebase (they wouldn't be right next to each other like they are here, so we'd want to be able to reason about their behavior orthogonally):
- `MaybeValid<T>`
- `try_cast_into<T>(bytes: &[u8]) -> Option<&MaybeValid<T>> where MaybeValid<T>: Freeze`
  - This is responsible for checking size and alignment, but not validity; in our codebase, it is used both by `FromBytes` (which needs no validity check) and by `TryFromBytes` (which performs a validity check after casting)
  - In our codebase (but not in this example), this does some complicated math to compute pointer metadata for slice DST; it can't easily be inlined into its caller
- `TryFromBytes`, with:
  -  `try_ref_from(bytes: &[u8]) -> Option<&Self> where Self: Freeze`
  -  `try_mut_from(bytes: &mut [u8]) -> Option<&mut Self>`

Notice the `MaybeValid<T>: Freeze` bound on `try_cast_into` (and the safety comments inside that function). That bound permits us to make the function safe to call (without that bound, we'd need a safety precondition about whether interior mutation ever happens using the argument/returned references).

`try_ref_from` can satisfy that bound because it already requires `Self: Freeze`, which means that `MaybeValid<Self>: Freeze`. However, `try_mut_from` does not require `Self: Freeze`. Today, the linked code fails to compile thanks to line 114 in `try_mut_from`:

```rust
let maybe_self = try_cast_into::<Self>(bytes)?;
```

What we'd like to do is be able to guarantee that interior mutation will never be exercised via `&MaybeValid<T>`, and thus be able to write `unsafe impl<T> Freeze for MaybeValid<T>`.

Reply:
> Notice the MaybeValid<T>: Freeze bound on try_cast_into (and the safety comments inside that function). That bound permits us to make the function safe to call (without that bound, we'd need a safety precondition about whether interior mutation ever happens using the argument/returned references).

What goes wrong if it's not Freeze? There can't even be a proof obligation attached to this since you want it to be trivially true. I think all you'd lose is some optimization potential.

If `MaybeValid` was always `Freeze`, you'd still have to extend the safety comments on `deref_unchecked` to say "also you must never ever mutate via this shared reference, even if `T` is e.g. `Cell<T>`".

Given that you need such a safety requirement anyway, MaybeValid is already a type that effectively permits no mutation of its interior ever. It then follows that you can drop the `Freeze` bound on `try_cast_into`. This requires `try_cast_into` to rely on a property of `MaybeValid`, but since `MaybeValid` appears in `try_cast_into`'s signature, there's no new coupling here -- you *already* depend on the fact that MaybeValid can be soundly constructed for invalid data.

So I still don't see a motivation for `impl Freeze` here, except if you somehow want more optimizations on `MaybeValid`.

Reply:
I see your point, and I think that's a compelling argument. I can imagine there being cases where you'd want to pass `MaybeValid` to a generic API that accepted `T: Freeze`, in which case you'd still need this, but I don't have a concrete example of this off the top of my head.

I'd still advocate for keeping this listed as a future possibility since doing so doesn't have any effect on the current proposal (ie, the current proposal is already forwards-compatible with this). Obviously we'd need to adjudicate it much more thoroughly if it was ever proposed to actually go through with permitting `unsafe impl Freeze`, and I assume you'd push back if that ever happened just as you are now, but I think it's at least worth mentioning that it's something that we've considered and that there might be some desire for.